crates/viewer/re_space_view_dataframe/src/dataframe_ui.rs

```diff
@@ -0,0 +1,350 @@
+        match self {
+            QueryHandle::LatestAt(query_handle) => query_handle.query().timeline,
+            QueryHandle::Range(query_handle) => query_handle.query().timeline,
+        }
+    }
+}
+
+impl<'a> From<LatestAtQueryHandle<'a>> for QueryHandle<'a> {
+    fn from(query_handle: LatestAtQueryHandle<'a>) -> Self {
+        QueryHandle::LatestAt(query_handle)
+    }
+}
+
+impl<'a> From<RangeQueryHandle<'a>> for QueryHandle<'a> {
+    fn from(query_handle: RangeQueryHandle<'a>) -> Self {
+        QueryHandle::Range(query_handle)
+    }
+}
+
+/// This structure maintains the data for displaying rows in a table.
+///
+/// Row data is stored in a bunch of [`DisplayRecordBatch`], which are created from
+/// [`RecordBatch`]s. We also maintain a mapping for each row number to the corresponding record
+/// batch and the index inside it.
+struct RowsDisplayData {
+    /// The [`DisplayRecordBatch`]s to display.
+    display_record_batches: Vec<DisplayRecordBatch>,
+
+    /// For each row to be displayed, the batch index and the row index with that batch.
+    batch_and_row_from_row_nr: BTreeMap<u64, (usize, usize)>,
+
+    /// The index of the time column corresponding to the query timeline.
+    time_column_index: Option<usize>,
+
+    /// The index of the time column corresponding the row IDs.
+    row_id_column_index: Option<usize>,
+}
+
+impl RowsDisplayData {
+    fn try_new(
+        row_indices: &Range<u64>,
+        record_batches: Vec<RecordBatch>,
+        schema: &[ColumnDescriptor],
+        query_timeline: &Timeline,
+    ) -> Result<Self, DisplayRecordBatchError> {
+        let display_record_batches = record_batches
+            .into_iter()
+            .map(|record_batch| DisplayRecordBatch::try_new(&record_batch, schema))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let mut batch_and_row_from_row_nr = BTreeMap::new();
+        let mut offset = row_indices.start;
+        for (batch_idx, batch) in display_record_batches.iter().enumerate() {
+            let batch_len = batch.num_rows() as u64;
+            for row_idx in 0..batch_len {
+                batch_and_row_from_row_nr.insert(offset + row_idx, (batch_idx, row_idx as usize));
+            }
+            offset += batch_len;
+        }
+
+        // find the time column
+        let time_column_index = schema
+            .iter()
+            .find_position(|desc| match desc {
+                ColumnDescriptor::Time(time_column_desc) => {
+                    &time_column_desc.timeline == query_timeline
+                }
+                _ => false,
+            })
+            .map(|(pos, _)| pos);
+
+        // find the row id column
+        let row_id_column_index = schema
+            .iter()
+            .find_position(|desc| match desc {
+                ColumnDescriptor::Control(control_column_desc) => {
+                    control_column_desc.component_name
+                        == ComponentName::from("rerun.controls.RowId")
```

Comment:
It would be nice if this wasn't quite as hard-coded

```suggestion
                        == ComponentName::row_id()
```